### PR TITLE
Enable SFTP protocol for scp(1) by default

### PR DESCRIPTION
**Pull request by djmdjm**
Enable SFTP for scp(1) by default. Implement scp -3 over SFTP.

### CODE REVIEW

File: sftp-client.h [hunk 1]
```diff
@@ -144,6 +144,21 @@ int do_upload(struct sftp_conn *, const char *, const char *, int, int, int);
 int upload_dir(struct sftp_conn *, const char *, const char *, int, int, int,
     int);
 
+/*
+ * Download a 'from_path' from the 'from' connection and upload it to
+ * to 'to' connection at 'to_path'.
+ */
+int do_crossload(struct sftp_conn *, struct sftp_conn *,
```
Comment (mfriedl) [inline sftp-client.h:151]: you need to keep the argument names for the comment

File: sftp-client.c [hunk 2]
```diff
@@ -2029,6 +2029,423 @@ upload_dir(struct sftp_conn *conn, const char *src, const char *dst,
...
+	if (!preserve_flag)
+		a->flags &= ~SSH2_FILEXFER_ATTR_ACMODTIME;
+	if (send_open(to, to_path, "dest",
+	    SSH2_FXF_WRITE|SSH2_FXF_CREAT|SSH2_FXF_TRUNC, a,
+	    &to_handle, &to_handle_len) != 0) {
+		do_close(from, from_handle, from_handle_len);
+		return -1;
+	}
+
+	/* Read from remote and write to local */
```
Comment (mfriedl) [inline sftp-client.c:2158]: write to remove

File: sftp-client.c [hunk 3]
```diff
@@ -2029,6 +2029,423 @@ upload_dir(struct sftp_conn *conn, const char *src, const char *dst,
...
+	/* Sanity check */
+	if (TAILQ_FIRST(&requests) != NULL)
+		fatal("Transfer complete, but requests still in queue");
+	/* Truncate at 0 length on interrupt or error to avoid holes at dest */
+	if (read_error || write_error || interrupted) {
+		debug("truncating \"%s\" at 0", to_path);
+		do_close(to, to_handle, to_handle_len);
+		if (send_open(to, to_path, "dest",
+		    SSH2_FXF_WRITE|SSH2_FXF_CREAT|SSH2_FXF_TRUNC, a,
+		    &from_handle, &from_handle_len) != 0) {
```
Comment (mfriedl) [inline sftp-client.c:2313]: to_handle

File: sftp-client.c [hunk 4]
```diff
@@ -2029,6 +2029,423 @@ upload_dir(struct sftp_conn *conn, const char *src, const char *dst,
...
+	if (read_error) {
+		error("Couldn't read from origin file \"%s\" : %s",
+		    from_path, fx2txt(status));
+		status = -1;
+		do_close(from, from_handle, from_handle_len);
+		do_close(to, to_handle, to_handle_len);
+	} else if (write_error) {
+		error("Couldn't write to \"%s\": %s",
+		    to_path, fx2txt(write_error));
+		status = SSH2_FX_FAILURE;
```
Comment (mfriedl) [inline sftp-client.c:2326]: since the status is not really used, you could set 'ret = 0/-1' instead

File: scp.c [hunk 5]
```diff
@@ -978,21 +988,58 @@ toremote(int argc, char **argv, enum scp_mode_e mode, char *sftp_direct)
...
+					    tport, sftp_direct,
+					    &remin, &remout, &do_cmd_pid);
+					if (conn == NULL) {
+						fatal("Unable to open "
+						    "destination connection");
+					}
+					debug3_f("origin in %d out %d pid %ld",
+					    remin, remout, (long)do_cmd_pid);
+				}
+				/* Connect to dest now */
```
Comment (mfriedl) [inline scp.c:1004]: source

File: scp.c [hunk 6]
```diff
@@ -978,21 +988,58 @@ toremote(int argc, char **argv, enum scp_mode_e mode, char *sftp_direct)
...
-			    *targ == '-' ? "-- " : "", targ);
-			if (do_cmd2(thost, tuser, tport, bp, remin, remout) < 0)
-				exit(1);
-			free(bp);
-			(void) close(remin);
-			(void) close(remout);
-			remin = remout = -1;
+			if (mode == MODE_SFTP) {
+				if (remin == -1) {
+					/* Connect to dest now */
```
Comment (mfriedl) [inline scp.c:993]: srrc or origin?

File: sftp-client.h (hunk 1 above)
Comment (djmdjm) [reply sftp-client.h]: fixed

File: sftp-client.c (hunk 2 above)
Comment (djmdjm) [reply sftp-client.c]: fixed

File: sftp-client.c (hunk 3 above)
Comment (djmdjm) [reply sftp-client.c]: oops - yes, there are also cases below where to_handle could be used after the send_open() failed that I have fixed too

Comment (djmdjm) [reply sftp-client.c]: (these weren't use-after-frees as send_open() sets the handle to NULL on error, but they would result in sending a garbage FXP_OPEN message)

File: sftp-client.c (hunk 4 above)
Comment (djmdjm) [reply sftp-client.c:2329]: done

File: scp.c (hunk 6 above)
Comment (djmdjm) [reply scp.c:993]: yes, origin

File: scp.c (hunk 5 above)
Comment (djmdjm) [reply scp.c]: fixed

File: scp.1 [hunk 7]
```diff
@@ -38,9 +37,21 @@ It uses
...
 for data transfer, and uses the same authentication and provides the
 same security as a login session.
-The scp protocol requires execution of the remote user's shell to perform
+Since OpenSSH 8.7,
+.Nm
+has used the SFTP protocol for file transfers and requires a working
+.Xr sftp-server 8
+(or equivalent) to be present on the server.
+Use of the legacy SCP protocol remains available via the
+.Fl O
```
Comment (pablomh) [inline scp.1:46]: Is this option missing in https://github.com/djmdjm/openssh-wip/pull/7/files#diff-b6c2513bd54d1f9dc161c39485138365b5d1f21f38f44e825e096bf8a911c4d2R21 an oversight or was it on purpose?

File: sftp-client.c [hunk 8]
```diff
@@ -2426,10 +2439,17 @@ crossload_dir_internal(struct sftp_conn *from, struct sftp_conn *to,
...
 			    depth + 1, &(dir_entries[i]->a), preserve_flag,
-			    print_flag) == -1)
+			    print_flag, follow_link_flag) == -1)
 				ret = -1;
-		} else if (S_ISREG(dir_entries[i]->a.perm) ) {
+		} else if (S_ISREG(dir_entries[i]->a.perm) ||
+		    (follow_link_flag && S_ISLNK(dir_entries[i]->a.perm))) {
+			/*
+			 * If this is a symlink then don't send the link's
+			 * Attrib. do_download() will do a FXP_STAT operation
```
Comment (mfriedl) [inline sftp-client.c:2448]: do_crossload ?

Comment (mfriedl) [reply sftp-client.c:2448]: no i'm wrong

File: scp.c [hunk 9]
```diff
@@ -1482,7 +1483,7 @@ sink_sftp(int argc, char *dst, const char *src, struct sftp_conn *conn)
 		debug("Fetching %s to %s\n", g.gl_pathv[i], abs_dst);
 		if (globpath_is_dir(g.gl_pathv[i]) && iamrecursive) {
 			if (download_dir(conn, g.gl_pathv[i], abs_dst, NULL,
-			    pflag, 1, 0, 0) == -1)
+			    pflag, SFTP_PROGRESS_ONLY, 0, 0) == -1)
```
Comment (mfriedl) [inline scp.c]: perhaps set this depending on the verbose flag? could be useful for debug.

File: scp.1 [hunk 10]
```diff
@@ -38,9 +37,21 @@ It uses
...
+option.
+.Pp
+The SCP protocol requires execution of the remote user's shell to perform
+.Xr glob 3
+pattern matching, whereas the SFTP protocol does not.
+This may result in some complex
 .Xr glob 3
-pattern matching.
+patterns being processed differently, depending on the protocol in use.
```
Comment (beldmit) [inline scp.1:54]: The other difference, according to the documentation, is the symlinks processing: SFTP doesn't follow them. SCP does...

File: scp.c (hunk 9 above)
Comment (djmdjm) [reply scp.c]: verbose already gives a lot more useful stuff since I hooked up log_init() in main :)

File: scp.1 (hunk 10 above)
Comment (djmdjm) [reply scp.1:54]: That is fixed in the last commit in this PR

Comment (djmdjm) [reply scp.1:54]: also, this is referring to the protocol and not the sftp(1) implementation

File: scp.1 (hunk 7 above)
Comment (djmdjm) [reply scp.1:46]: good catch. -O is missing from the SYNOPSIS section. I'll add it in the next patchset

**Pull request closed without merging**